### DISCUSSION TITLE
Right commands for P7 -> P8 conversion MKV

### DISCUSSION THREAD

**Discussion by hyperknot** [quietvoid/dovi_tool#290] [General]
Hi,

I'm trying to figure out the right commands to convert P7 -> P8 on MKV files.

I'm running macOS, so I've invested quite some time in trying to understand what the Windows only scripts are doing.

[DoVi_Scripts](https://github.com/R3S3t9999/DoVi_Scripts) does the following:
```
ffmpeg -i $MKVFILE -c:v copy -bsf:v hevc_mp4toannexb -f hevc - | dovi_tool demux -

dovi_tool -m 2 mux --bl BL.hevc --el EL.hevc --discard -o P8.hevc

mkvmerge --output $MKVFILE-P8.mkv --compression 0:none --default-duration 0:24000/1001p P8.hevc --no-video --compression 1:none $MKVFILE
```

Here I believe the last command could be simplified to:
```
mkvmerge --output $MKVFILE-P8mod.mkv P8.hevc --no-video $MKVFILE
```

[nekno DV7toDV8](https://github.com/nekno/DV7toDV8) does the following:
```
mkvextract "$mkvFile" tracks 0:"$BL_EL_RPU_HEVC"
dovi_tool demux --el-only "$BL_EL_RPU_HEVC" -e "$DV7_EL_RPU_HEVC"
dovi_tool --edit-config "$jsonFile convert --discard "$BL_EL_RPU_HEVC" -o "$DV8_BL_RPU_HEVC"
dovi_tool extract-rpu "$DV8_BL_RPU_HEVC" -o "$DV8_RPU_BIN"
mkvmerge -o "$mkvBase.DV8.mkv" -D "$mkvFile" "$DV8_BL_RPU_HEVC" --track-order 1:0
```

There is also [DDVT tool](https://forum.doom9.org/showthread.php?t=183479) and [this snippet
](https://www.reddit.com/r/ffmpeg/comments/11gu4o4/comment/jn5gman/) in Reddit.

```
ffmpeg -y -i 'path_to.mkv' -dn -c:v copy -vbsf hevc_mp4toannexb -f hevc - | dovi_tool -m 2 convert --discard - -o 'path_to.hevc'
```

Can you tell me which of them are right? Also, couldn't `dovi_tool demux -` plus `dovi_tool -m 2 mux` be simplified into a single command in the first example?

**Comment by quietvoid** [comment]
In theory `dovi_tool -m 2 convert --discard` is sufficient.
It assumes that BL/EL are in sync but very rarely they are not, in which case it's best to use `extract-rpu` first then `inject-rpu`.

> Also, couldn't dovi_tool demux - plus dovi_tool -m 2 mux be simplified into a single command in the first example?

Yes.

There were issues in the past with `ffmpeg` causing the output stream to have decode errors, so sometimes `mkvextract` may be more reliable.

**Reply by hyperknot** [reply]
Thanks! How do I find out if they are not in sync? Will it print something? Or it'll be just artifacts when playing back the file on TV?

**Reply by hyperknot** [reply]
Can you tell me the correct command for mkvextract to substitute this? Or it's doing this automatically?
```
ffmpeg -i $MKVFILE -c:v copy -bsf:v hevc_mp4toannexb -f hevc -
```

**Reply by quietvoid** [reply]
There wouldn't be any indication, you'd have to manually check the frames' PTS are identical.
The specification requires them to be, anyways.

One automated way would be to do multiple operations to verify that the output RPU is the same:
```console
dovi_tool demux video.hevc --el-only

# BL PTS output
dovi_tool -m 0 extract-rpu video.hevc -o BL_RPU.bin

# EL PTS output
dovi_tool -m 0 extract-rpu EL.hevc -o EL_RPU.bin

# Should output the same checksum
sha512sum BL_RPU.bin EL_RPU.bin
```

**Reply by quietvoid** [reply]
> Can you tell me the correct command for mkvextract to substitute this? Or it's doing this automatically?
> 
> ```
> ffmpeg -i $MKVFILE -c:v copy -bsf:v hevc_mp4toannexb -f hevc -
> ```

That's equivalent to `mkvextract $MKVFILE tracks 0:video.hevc`, assuming the video has ID 0.

**Reply by hyperknot** [reply]
Thanks, I've done:

```
ffmpeg -i $MKVFILE -c:v copy -bsf:v hevc_mp4toannexb -f hevc - > outf
mkvextract $MKVFILE tracks 0:outm
```

And they are slightly different:
```
ls -al out*
-rw-r--r--  1 user  staff  683156522 Mar  6 19:27 outf
-rw-r--r--  1 user  staff  683095436 Mar  6 19:27 outm
```

Should I be concerned? Mediainfo is identical, except for the size.

**Reply by quietvoid** [reply]
Not really, it could just be a difference in NAL headers size.
Both are safe to use 99.9% of the time.

**Reply by hyperknot** [reply]
Thanks a lot!

Here is my full snippet:

```
rm -f -- *.hevc

mkvextract "$1" tracks 0:video.hevc

# check if BL/EL are in sync
dovi_tool demux video.hevc --el-only
dovi_tool -m 0 extract-rpu video.hevc -o BL_RPU.bin
dovi_tool -m 0 extract-rpu EL.hevc -o EL_RPU.bin

if cmp -s BL_RPU.bin EL_RPU.bin; then
  echo "BL/EL are in sync"
  rm BL_RPU.bin EL_RPU.bin
else
  echo "BL/EL are not in sync"
  rm BL_RPU.bin EL_RPU.bin
  exit 1
fi

dovi_tool -m 2 convert --discard video.hevc -o P8.hevc

mkvmerge --output "$1-P8.mkv" P8.hevc --no-video "$1"

rm -f -- *.hevc
```

This means I can get rid of all the virtualized windows with is slow emulated network and just run this on Linux / macOS.
